stackrox/stackrox
# ROX-18609: Introduce ClusterHealthResponse

## Description

This PR adds a response to the `RawClusterHealthInfo` message from Sensor, establishing constant writes from `Central->Sensor` in the interval of `RawClusterHealthInfo` messages being sent (currently 30s).

The reason for this is when Central is exposed behind an Ingress controller that does TLS termination, our current way of [keepalives via gRPC](https://github.com/grpc/grpc/blob/master/doc/keepalive.md) isn't working as expected. This is due to the ingress controller not forwarding the HTTP/2 frame (as a reference, here's [an issue for NGINX explaining this](https://trac.nginx.org/nginx/ticket/1887)).
This leads to infrequent writes from `Central->Sensor` (the latest will be the reprocessing interval duration, i.e. 4 hours). This then triggers `write timeouts` defined within the ingress controller.

Currently, the only way to deal with this is to increase the timeout to an increased value (like the aforementioned 4 hours). This is also suggested by [the NGINX ingress controller documentation for long-running bi-directional streams.](https://kubernetes.github.io/ingress-nginx/examples/grpc/#notes-on-using-responserequest-streams)

As suggested[ in the issue opened for NGINX](https://trac.nginx.org/nginx/ticket/1887), an alternative is to introduce health checks / constant flow of messages in the application, which this PR does by responding to the frequently sent cluster health messages.

As a note for reviewers, this PR superseeds #7161. The introduction of central capabilities will be split into a different PR.

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

- see CI (especially GKE compatibility tests)

Manual tests (for the test environment Kong was used as ingress controller of choice, you probably can use any other ingress controller):

1. Create an EKS cluster on infra.

2. Have a secondary K8S cluster (this can be either local, OS, or GKE).

_Below steps are all executed on the EKS cluster_

3. Install Kong following [the documentation here for EKS](https://docs.konghq.com/kubernetes-ingress-controller/latest/deployment/eks/). Alternatively, you can use any other ingress controller, but have to adjust the below manifest and steps accordingly.
Also, do not forget to create the ingress class:
```bash
echo "
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  name: kong
spec:
  controller: ingress-controllers.konghq.com/kong
" | kubectl apply -f -
```

5. For the Kong service, ensure that the load balancer type `ELB` is used.
```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
    service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "80"
    service.beta.kubernetes.io/aws-load-balancer-type: elb
```

6. Associate a public domain with the load balancer public IP and create certificates for it e.g. via `certbot`:
```bash
sudo certbot certonly --manual -d <your-domain-name> --preferred-challenges dns
```

7. Create a secret within the `stackrox` namespace with the generated cert/key pair from `certbot`:
```yaml
kubectl create ns stackrox
kubectl create secret tls stackrox-rsa-custom -n stackrox --cert=fullchain.pem --key=privkey.pem
```

8. Generate a Central bundle via `roxctl`:
```bash
# PSP = false
# Expose = none (we will create a custom Ingress for this)
roxctl central generate interactive
```

9. Adjust the config map manifest containing the endpoint config (`01-central-09-endpoints-config.yaml`):
```yaml
...
data:
  endpoints.yaml: |
    endpoints:
      - listen: "8080"
        tls:
          disable: true
```

10. Adjust the central service manifest to include the 8080 port and the kong annotation(`01-central-14-service.yaml`):
```yaml
...
  annotations:
    konghq.com/protocol: "grpc"
...
spec:
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 8080
```

11. Deploy the central by running:
```yaml
./central-bundle/central/scripts/setup.sh
kubectl create -R -f central-bundle/central
```

12. Afterwards, create an ingress that exposes the 8080 port with TLS terminated using the previously created `stackrox-rsa-custom` secret:
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    konghq.com/protocols: grpc,grpcs,https,http
    kubernetes.io/ingress.class: kong
  name: central-ingress
  namespace: stackrox
spec:
  rules:
  - host: <custom domain name>
    http:
      paths:
      - backend:
          service:
            name: central
            port:
              number: 80
        path: /
        pathType: ImplementationSpecific
  tls:
  - hosts:
    - <custom domain name>
    secretName: stackrox-rsa-custom
```

13. Verify that the UI opens successfully and you can connect to your Central using `https://<custom domain name>`.

_Below commands are all executed on the secondary K8S cluster (OpenShift, local, GKE)_
14. Deploy the sensor within the cluster:
```bash
export API_ENDPOINT="https://<custom domain name>:443"
export ROX_ADMIN_PASSWORD=$(cat central-bundle/password)
./deploy/sensor.sh
kubectl set env -n stackrox deploy sensor ROX_CENTRAL_ENDPOINT=<custom domain name>:443
```

15. Verify that sensor is successfully connecting to Central and shown as healthy.

16. Observe that the connection is not interrupted between Sensor and Central. You can determine this by the pod restarts of Sensor as well as observing the logs of Central. On local clusters with little to no workload, the pod previously restarted after latest 10 minutes. On OpenShift or GKE clusters with more workloads, this usually takes not longer than 30 minutes.
